src/Payments.sol

```diff
@@ -0,0 +1,572 @@
+        returns (
+            uint256 totalSettledAmount,
+            uint256 finalSettledEpoch
+        )
+    {
+        // Get the rail and the involved accounts.
+        Rail storage rail = rails[railId];
+        Account storage payer = accounts[rail.token][rail.from];
+        Account storage payee = accounts[rail.token][rail.to];
+
+        settleAccountLockup(payer);
+
+        uint256 settlementTargetEpoch = min(untilEpoch, payer.lockupLastSettledAt + rail.lockupPeriod);
+
+        // Begin settlement from the last settled epoch.
+        uint256 currentSettlementEpoch = rail.lastSettledAt;
+        uint256 currentRate = rail.paymentRate;
+        totalSettledAmount = 0;
+
+        // Use the rail's rate–change queue.
+        RateChangeQueue.Queue storage rateQueue = railRateChangeQueues[railId];
+
+        // If there are no queued rate changes, settle the entire segment in one shot.
+        if (rateQueue.isEmpty()) {
+            (totalSettledAmount, finalSettledEpoch) = _settleSegment(
+                railId,
+                rail,
+                currentSettlementEpoch,
+                settlementTargetEpoch,
+                currentRate
+            );
+        } else {
+            // Otherwise, settle the rail in segments (each up to the next rate–change boundary).
+            (totalSettledAmount, finalSettledEpoch) = _settleWithRateChanges(
+                railId,
+                rail,
+                rateQueue,
+                currentRate,
+                currentSettlementEpoch,
+                settlementTargetEpoch
+            );
+        }
+
+        // Update account balances.
+        payer.funds -= totalSettledAmount;
+        payee.funds += totalSettledAmount;
+        require(payer.lockupCurrent >= totalSettledAmount, "Insufficient lockup funds");
+        payer.lockupCurrent -= totalSettledAmount;
+
+        // Record the new settlement epoch on the rail.
+        rail.lastSettledAt = finalSettledEpoch;
+
+        require(payer.lockupCurrent <= payer.funds, "insufficient funds");
+
+        return (totalSettledAmount, finalSettledEpoch);
+    }
+
+    function _settleSegment(
+        uint256 railId,
+        Rail storage rail,
+        uint256 segmentStart,
+        uint256 segmentEnd,
+        uint256 rate
+    ) internal returns (uint256 settledAmount, uint256 settledUntil) {
+        // Calculate the intended payment for the entire segment.
+        uint256 duration = segmentEnd - segmentStart;
+        settledAmount = rate * duration;
+        settledUntil = segmentEnd;
+
+        if (rail.arbiter != address(0)) {
+            // Call the external arbitrator.
+            IArbiter arbiter = IArbiter(rail.arbiter);
+            IArbiter.ArbitrationResult memory arbResult = arbiter.arbitratePayment(
+                railId,
+                settledAmount,
+                segmentStart,
+                segmentEnd
+            );
+            require(arbResult.approved, "arbitrer refused payment");
```

Comment:
We need to validate that:

1. `arbResult.settledUntil <= settledUntil`.
2. `arbResult.modifiedAmount <= rate * (arbResult.settledUntil - segmentStart)`.

Reply:
For 1, it should be `arbResult.settledUntil <=segmentEnd`, right ? Yeah making this change.

Reply:
Done.